File: java/Zoo.java
import java.util.Comparator;
import java.util.TreeMap;
import java.util.Map;

public class Zoo {
    private Map<String, Animal> animals;

    public Zoo() {
        animals = new TreeMap<>();
    }

    public Animal add(Animal animal) {
        Animal replaced = animals.put(animal.getName(), animal);
        return replaced;
    }

    public Animal remove(String name) {
        return animals.remove(name);
    }

    public void print() {
        for (Animal animal : animals.values()) {
            System.out.println(animal);
        }
    }

    public void printSortedByName() {
        for (Animal animal : animals.values()) {
            System.out.println("Key: [" + animal.getName() +
                    "], Value: [" + animal + "]");
        }
    }

    // 5. Add a method in Zoo to find animals by name.
    public Animal findByName(String name) {
        return animals.get(name);
    }

    public void printSortedByAge() {
        animals.values().stream()
                .sorted(Comparator.comparing(Animal::getBirth)
                        .thenComparing(Animal::getWeight)
                        .thenComparing(Animal::getName))
                .forEach(System.out::println);
    }
}

File: cpp/Zoo.cpp
#include "Zoo.h"

#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

#include "Animal.h"
#include "Date.h"
#include "Tiger.h"

Animal *Zoo::add(Animal *a) {
  Animal *replaced = nullptr;
  auto it = animals.find(a->getName());
  if (it != animals.end()) {
    replaced = it->second;
  }
  animals[a->getName()] = a;
  return replaced;
}

Animal *Zoo::remove(const std::string &name) {
  auto it = animals.find(name);
  if (it != animals.end()) {
    Animal *removed = it->second;
    animals.erase(it);
    return removed;
  }
  return nullptr;
}

void Zoo::print() {
  for (const auto &pair : animals) {
    std::cout << pair.second->ToString() << std::endl;
  }
}

void Zoo::printSortedByName() {
  for (const auto &pair : animals) {
    std::cout << "Key: [" << pair.first << "], Value: ["
              << pair.second->ToString() << "]" << std::endl;
  }
}

void Zoo::printSortedByAge() {
  // Date sisältää operator< järjestämistä varten
  // Multimap mahdollistaa usean eläimen samalla syntymäpäivällä
  std::multimap<Date, Animal *> sortedAnimals;
  for (const auto &pair : animals) {
    sortedAnimals.insert(std::make_pair(pair.second->getBirth(), pair.second));
  }
  for (const auto &pair : sortedAnimals) {
    std::cout << pair.second->ToString() << std::endl;
  }
}

Zoo::~Zoo() {
  for (auto &pair : animals) {
    delete pair.second;
  }
  animals.clear();
}